doc/guides/dts/index.rst

```diff
@@ -346,44 +346,534 @@ Below is sample header content generated for this node, in
    #define DT_INST_0_NXP_KINETIS_SIM_LABEL             DT_NXP_KINETIS_SIM_40047000_LABEL
    #define DT_INST_0_NXP_KINETIS_SIM                   1
 
-Generated macro names follow the format ``DT_<node>_<property>``. For
-``DT_NXP_KINETIS_SIM_40047000_BASE_ADDRESS``, the node part is
-``NXP_KINETIS_SIM_40047000``, based on the compatible string that matched a
-binding for the node (``nxp,kinetis-sim``) and the node's unit address
-(``...@4004700``).
+All macros generated from devicetree start with ``DT_`` and use all-uppercase.
+Most macro names contain a node identifier, and macros that correspond to
+devicetree properties also contain an identifier for the property.
 
-The ``*_BASE_ADDRESS`` part of the identifier is a fixed identifier generated
-from the special ``reg`` property on the node. ``*_SIZE`` is also generated
-from ``reg``. Other suffixes, like ``*_LABEL``, are generated directly from the
-property name.
+Node identifiers in macro names
+-------------------------------
 
-The second macro (``DT_INST_0_NXP_KINETIS_SIM_BASE_ADDRESS``) is an alias for
-the first macro. The node identifier ``...INST_0_NXP_KINETIS_SIM_...`` means
-"the first node with compatible string ``nxp,kinetis-sim``.
+Macros generated from particular devicetree nodes start with ``DT_<node>``,
+where ``<node>`` identifies the node. Several identifiers are generated for
+nodes:
 
-Aliases are also generated from any properties in the ``/aliases`` node. Take
-the DTS fragment below as an example.
+``DT_(<bus>_)<compatible>_<unit-address>``
+    The compatible string for the node followed by its unit address. If the
+    node has several compatible strings, the first one that matched a binding
+    is used.
+
+    If the node appears on a bus (has ``on-bus:`` in its binding), then the
+    compatible string and unit address of the bus node is put before the
+    compatible string for the node itself.
+
+    For nodes that don't have a unit address (no ``...@<address>`` in the node
+    name), the unit address of the parent node is used instead. If the parent
+    node has no unit address either, the name of the node is used for
+    ``<unit-address>``, as a fallback.
+
+    For the ``sim@40047000`` node above, the compatible string is
+    ``"nxp,kinetis-sim"``. This gets converted to ``NXP_KINETIS_SIM`` by
+    uppercasing and replacing non-alphanumeric characters with underscores.
+    Adding the unit address gives the node prefix
+    ``DT_NXP_KINETIS_SIM_40047000``.
+
+``DT_INST_<instance-no.>_<compatible>``
+    The compatible string for the node together with an instance number.
+
+    The instance number is a unique index among all enabled
```

Comment:
Please add a note that this does not in any way match the datasheet instance, and that there is no guarantee that two builds will yield the same instance numbers (which I find weird btw, but OK).

Reply:
Here's what I'm adding:

```
       The instance numbers in no way reflect any numbering scheme that
       might exist for the SoC.  There is no guarantee that the same device
       node is given the same instance number between builds.  The only
       guarantee is that instance numbers will start at 0 and be contingous
       for the number of compatiable instances that are enabled.
```

Reply:
One more thing @galak, I would add that it is also completely unrelated to numbers appearing in the the node label or node name. For example:
```
&uart0 {
	compatible = "nordic,nrf-uart";
	status = "okay";
	current-speed = <115200>;
	tx-pin = <6>;
	rx-pin = <8>;
	rts-pin = <5>;
	cts-pin = <7>;
};
```

there is no guarantee that this will end up as `DT_INST_0`

Reply:
Changed to:

```
       The instance numbers in no way reflect any numbering scheme that
       might exist for the SoC, in any node labels, or unit addresses for the 
       node.  The instance number is a pure count of enabled nodes with the
       same compatible.  There is no guarantee that the same device
       node is given the same instance number between builds.  The only
       guarantee is that instance numbers will start at 0 and be contiguous
       for the number of compatible instances that are enabled.
```